File: lib_book_common/src/main/java/com/ebook/common/adapter/BaseBindAdapter.java
package com.ebook.common.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ObservableArrayList;
import androidx.databinding.ViewDataBinding;
import androidx.recyclerview.widget.RecyclerView;


public abstract class BaseBindAdapter<T, B extends ViewDataBinding> extends RecyclerView.Adapter {
    protected Context context;
    protected ObservableArrayList<T> items;
    protected OnItemClickListener<T> mItemClickListener;
    protected OnItemLongClickListener<T> mOnItemLongClickListener;

    public BaseBindAdapter(Context context, ObservableArrayList<T> items) {
        this.context = context;
        this.items = items;
    }

    @Override
    public int getItemCount() {
        return items != null && items.size() > 0 ? items.size() : 0;
    }

    @Override
    @NonNull
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        B binding = DataBindingUtil.inflate(LayoutInflater.from(this.context), this.getLayoutItemId(viewType), parent, false);
        return new BaseBindingViewHolder(binding.getRoot());
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        B binding = DataBindingUtil.getBinding(holder.itemView);
        this.onBindItem(binding, this.items.get(position), position);
    }

    @LayoutRes
    protected abstract int getLayoutItemId(int viewType);

    protected abstract void onBindItem(B binding, T item, int position);

    public void setItemClickListener(OnItemClickListener<T> itemClickListener) {
        mItemClickListener = itemClickListener;
    }

    public void setOnItemLongClickListener(OnItemLongClickListener<T> onItemLongClickListener) {
        mOnItemLongClickListener = onItemLongClickListener;
    }

    public interface OnItemClickListener<E> {
        void onItemClick(E e, int position);
    }

    public interface OnItemLongClickListener<E> {
        boolean onItemLongClick(E e, int postion);
    }

    public static class BaseBindingViewHolder extends RecyclerView.ViewHolder {
        BaseBindingViewHolder(View itemView) {
            super(itemView);
        }
    }

}

File: lib_common/src/main/java/com/xrn1997/common/adapter/BaseBindAdapter.kt
package com.xrn1997.common.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.databinding.DataBindingUtil
import androidx.databinding.ObservableArrayList
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.RecyclerView

/**
 * 使用DataBinding的BaseAdapter
 * @author xrn1997
 */
@Suppress("unused")
abstract class BaseBindAdapter<T, B : ViewDataBinding>(
    @JvmField
    protected var context: Context,
    protected var items: ObservableArrayList<T>
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    @JvmField
    protected var mOnItemClickListener: OnItemClickListener<T>? = null

    @JvmField
    protected var mOnItemLongClickListener: OnItemLongClickListener<T>? = null
    override fun getItemCount(): Int {
        return if (items.size > 0) items.size else 0
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding: B = DataBindingUtil.inflate(
            LayoutInflater.from(context), getLayoutItemId(
                viewType
            ), parent, false
        )
        return BaseBindingViewHolder(binding.root)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val binding: B? = DataBindingUtil.getBinding(holder.itemView)
        items[position]?.let { onBindItem(binding!!, it, position) }
    }

    class BaseBindingViewHolder internal constructor(itemView: View?) : RecyclerView.ViewHolder(
        itemView!!
    )

    /**
     * 获得item的布局id
     */
    @LayoutRes
    protected abstract fun getLayoutItemId(viewType: Int): Int

    protected abstract fun onBindItem(binding: B, item: T, position: Int)

    /**
     * item监听
     */
    open fun setOnItemClickListener(itemClickListener: OnItemClickListener<T>) {
        mOnItemClickListener = itemClickListener
    }

    /**
     * item长按监听
     */
    open fun setOnItemLongClickListener(onItemLongClickListener: OnItemLongClickListener<T>?) {
        mOnItemLongClickListener = onItemLongClickListener
    }

    /**
     * item click监听接口
     */
    interface OnItemClickListener<E> {
        fun onItemClick(e: E, position: Int)
    }

    /**
     * item long click监听接口
     */
    interface OnItemLongClickListener<E> {
        fun onItemLongClick(e: E, position: Int): Boolean
    }

}